src/main.rs

```diff
@@ -105,21 +105,61 @@ fn determine_kind<'a>(cx: &'a Context<'_>, id: &PackageId, progress: &mut Progre
             .iter()
             .map(String::as_str)
             .filter(|f| *f != "default" && !cx.exclude_features.contains(f))
+            .map(Cow::Borrowed)
             .collect();
+
         if let Some(opt_deps) = &cx.optional_deps {
-            features.extend(package.dependencies.iter().filter_map(Dependency::as_feature).filter(
-                move |&f| {
-                    !cx.exclude_features.contains(&f)
-                        && (opt_deps.is_empty() || opt_deps.contains(&f))
-                },
-            ));
+            features.extend(
+                package
+                    .dependencies
+                    .iter()
+                    .filter_map(Dependency::as_feature)
+                    .filter(move |f| {
+                        !cx.exclude_features.contains(f)
+                            && (opt_deps.is_empty() || opt_deps.contains(f))
+                    })
+                    .map(Cow::Borrowed),
+            );
         }
+
+        if cx.include_deps_features {
+            let node = cx.nodes(id);
+            let package = cx.packages(id);
+            // TODO: Unpublished dependencies are not included in `node.deps`.
+            for dep in node.deps.iter().filter(|dep| {
+                // ignore if `dep_kinds` is empty (i.e., not Rust 1.41+), target specific or not a normal dependency.
+                dep.dep_kinds.iter().any(|kind| kind.kind.is_none() && kind.target.is_none())
+            }) {
+                let dep_package = cx.packages(&dep.pkg);
+                let dep_node = cx.nodes(&dep.pkg);
+                features.extend(dep_package.features.iter().filter_map(|f| {
+                    // ignore if features already enabled on the crate
+                    if !dep_node.features.contains(f) {
```

Comment:
Hmm, this seems not good. IIUC, as `resolve.nodes[].features` is based on the feature flags when `cargo metadata` was run, it includes default features and features enabled by dev-dependencies...